presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushProjectionThroughExchange.java

```diff
@@ -118,17 +116,23 @@ public Result apply(ProjectNode project, Captures captures, Context context)
                         .filter(symbol -> !partitioningColumns.contains(symbol))
                         .map(outputToInputMap::get)
                         .forEach(nameReference -> {
-                            Symbol symbol = Symbol.from(nameReference);
-                            projections.put(symbol, castToRowExpression(nameReference));
-                            inputs.add(symbol);
+                            projections.put(nameReference, castToRowExpression(nameReference.toSymbolReference()));
+                            inputs.add(nameReference);
                         });
             }
 
             for (Map.Entry<Symbol, RowExpression> projection : project.getAssignments().entrySet()) {
-                Expression translatedExpression = inlineSymbols(outputToInputMap, castToExpression(projection.getValue()));
+                RowExpression translatedExpression = RowExpressionSymbolInliner.inlineSymbols(outputToInputMap, projection.getValue());
                 Type type = context.getSymbolAllocator().getTypes().get(projection.getKey());
-                Symbol symbol = context.getSymbolAllocator().newSymbol(translatedExpression, type);
-                projections.put(symbol, castToRowExpression(translatedExpression));
+                Symbol symbol;
+                if (isExpression(translatedExpression)) {
```

Comment:
An optimizer rule should only handle `Expression`; otherwise it's bug. remove the `else` branch.

Reply:
This commit might not be named correctly. It is update mainly because the test was updated to use RowExpression. Given we are migrating Expression to RowExpression. It should be only beneficial if we handle RowExpression as well here?

Reply:
Then use `castToRowExpression` in the test. Check `TestEffectivePredicateExtractor`